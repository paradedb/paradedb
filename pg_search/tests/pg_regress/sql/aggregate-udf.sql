CREATE EXTENSION IF NOT EXISTS pg_search;
DROP TABLE IF EXISTS pr2625;
CREATE TABLE pr2625 (
    id serial8,
    k text,
    v float8
);
CREATE INDEX idxpr2625 ON pr2625 USING bm25 (id, k, v) WITH (key_field = 'id', target_segment_count = 8, layer_sizes = '1gb, 10gb');

--
-- test with 1 segment
--
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;

SET parallel_leader_participation = true;
SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>true);
SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>false);
SET parallel_leader_participation = false;
SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>true);


--
-- test with multiple segments
--
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;

SET parallel_leader_participation = true;
SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>true);
SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>false);
SET parallel_leader_participation = false;
SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>true);


--
-- this one will execute serially in the current backend
--
SET parallel_leader_participation = false;
SET max_parallel_workers = 0;
SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>true);

RESET parallel_leader_participation;
RESET max_parallel_workers;
