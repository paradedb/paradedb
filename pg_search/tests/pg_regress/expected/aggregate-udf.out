DROP TABLE IF EXISTS pr2625;
CREATE TABLE pr2625 (
    id serial8,
    k text,
    v float8
);
CREATE INDEX idxpr2625 ON pr2625 USING bm25 (id, k, v) WITH (key_field = 'id');
--
-- test with 1 segment
--
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;
SET parallel_leader_participation = true;
SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>true);
           aggregate           
-------------------------------
 {"average": {"value": 500.5}}
(1 row)

SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>false);
           aggregate           
-------------------------------
 {"average": {"value": 500.5}}
(1 row)

SET parallel_leader_participation = false;
SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>true);
           aggregate           
-------------------------------
 {"average": {"value": 500.5}}
(1 row)

--
-- test with multiple segments
--
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;
INSERT INTO pr2625(k, v) SELECT paradedb.random_words(1), x::float8 from generate_series(1, 1000) x;
SET parallel_leader_participation = true;
SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>true);
           aggregate           
-------------------------------
 {"average": {"value": 500.5}}
(1 row)

SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>false);
           aggregate           
-------------------------------
 {"average": {"value": 500.5}}
(1 row)

SET parallel_leader_participation = false;
SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>true);
           aggregate           
-------------------------------
 {"average": {"value": 500.5}}
(1 row)

--
-- this one should generate an ERROR
--
SET parallel_leader_participation = false;
SET max_parallel_workers = 0;
SELECT * FROM paradedb.aggregate(index=>'idxpr2625', query=>paradedb.all(), agg=>'{"average": { "avg": { "field": "v" }}}', solve_mvcc=>true);
ERROR:  should be able to launch parallel process
RESET parallel_leader_participation;
RESET max_parallel_workers;
